integrations/electricity/energinet.js

```diff
@@ -0,0 +1,184 @@
+  {
+    datetime: String # ISO datetime format,
+    value: Float,
+    mRID: String # a use might have several meters associated with him
+  }
+*/
+async function getTimeSeries(accessToken, meterPointIds, lastCollect) {
+  const now = moment();
+  const dateFrom = lastCollect.clone();
+  if (
+    dateFrom
+      .clone()
+      .add(1, 'hour')
+      .isAfter(now)
+  ) {
+    return [];
+  }
+  const dateTo = moment.min(lastCollect.clone().add(14, 'days'), moment());
+  const url = TIME_SERIES_URL.replace('{dateFrom}', dateFrom.format(DATE_FORMAT))
+    .replace('{dateTo}', dateTo.format(DATE_FORMAT))
+    .replace('{aggregation}', AGGREGATION);
+  const bodyMeterPoints = {
+    meteringPoints: {
+      meteringPoint: meterPointIds,
+    },
+  };
+  const res = await request
+    .post(url)
+    .send(bodyMeterPoints)
+    .set('Authorization', `Bearer ${accessToken}`)
+    .set('Content-Type', 'application/json');
+  if (!res.ok) {
+    throw new HTTPError(res.text, res.status);
+  }
+
+  const timeSeries = res.body.result[0].MyEnergyData_MarketDocument.TimeSeries.map(
+    meteringPoint => {
+      const { mRID } = meteringPoint;
+      return meteringPoint.Period.map(period => {
+        const periodStart = period.timeInterval.start;
+        return period.Point.map(periodPoint => {
+          return {
+            datetime: moment(periodStart)
+              .clone()
+              .add(parseInt(periodPoint.position, 10) - 1, 'hours')
+              .toISOString(),
+            value: parseFloat(periodPoint['out_Quantity.quantity']),
+            mRID,
+          };
+        });
+      });
+    }
+  ).flat(Infinity);
+
+  return timeSeries.concat(await getTimeSeries(accessToken, meterPointIds, dateTo));
+}
+
+async function connect({ requestToken }, logger) {
+  const { token } = await requestToken();
+  // Test that the provided token is valid and allows to fetch an access token
+  // Test of all requests used for collect
+  const accessToken = await getAccessToken(token);
+  const { meterPointIds, meterPointAddresses } = await getMeteringPoints(accessToken);
+  const timeSeries = await getTimeSeries(accessToken, meterPointIds, moment().subtract(3, 'days'));
+
+  // Take first meter point as reference for address
+  let lonLat = [null, null];
+  if (meterPointAddresses.length > 0) {
+    lonLat = await cityToLonLat('DK', meterPointAddresses[0].postcode);
+  }
+
+  return {
+    refreshToken: token,
+    locationLon: lonLat[0],
+    locationLat: lonLat[1],
+  };
+}
+
+async function collect(state, logger) {
+  const { refreshToken, locationLon, locationLat } = state;
+  const accessToken = await getAccessToken(refreshToken);
+  const { meterPointIds, meterPointAddresses } = await getMeteringPoints(accessToken);
+
+  // Fetch from last update. If not available, then fetch data from the last 90 days.
+  const lastCollect = state.lastCollect ? moment(state.lastCollect) : moment().subtract(90, 'days');
```

Comment:
What about fetching for the last year so it can be used to make the carbon accounting for a full year?

Reply:
We could change that yes, again I used the orsted integration as a template. I am not sure it's really interesting for north, but for Bloom it could indeed have value

Reply:
Let's fetch as much as possible IMO (as two years would also allow us to do YoY changes)

Reply:
Let's go with one year then, test it, and the amount of data is not already overwhelming for the device, let's bump it to two years